### PR TITLE
test: fs: Test cases for file system interfaces

### PR DESCRIPTION
**Pull request by mengxianglinx**
Focus on file system interfaces implemented in file system core,
not specific file system.

Signed-off-by: Meng xianglin <xianglinx.meng@intel.com>

### CODE REVIEW

**Review by pabigot — COMMENTED** [review]
What's the motivation for this?

For something like `timer_api` we have a standard set of test cases that are run against each implementation of the API.

Here it looks like you're not testing the real implementations, only a stub implementation that's within the test itself.

Do you intend to add infrastructure to test specific instances of file systems, e.g. have overlays that independently enable FatFs or littlefs on specific flash partitions (possibly not located on a flash device) to test those implementations?

**Comment by mengxianglinx** [comment]
Hi @pabigot , thank you for your question.
This patch aims to  test the infrastructure of file system which is mainly implemented in subsys/fs/fs.c, test the behavior of interfaces such as fs_open fs_readdir in fs.c which in turn call interfaces implemented in specific file system. Existing test only test apis in specific file system.
Since there is no need to test specific file system but this test really need a specific file system, I disable FatFs and littlefs and include a stub implementation for the sake of simplicity.

File: tests/subsys/fs/fs_api/src/test_multi_fs.c [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+	if (fs_unmount(&test_fs_mnt_2)) {
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int test_fs_unsupported(void)
+{
+	if (fs_register(UNSUPPORTED_FS, &temp_fs) == 0) {
```
Comment (pabigot) [inline tests/subsys/fs/fs_api/src/test_multi_fs.c:92]: I think you should be using `FS_TYPE_END` here explicitly, and in the `.type` initialization of `test_fs_mount_unsupported_fs`.  It's that value, representing the maximum number of file systems, that isn't allowed.

This reveals a weakness in the common code.  `enum fs_type` provides "unique" identifiers for file systems.  It should allow for externally-defined file systems and it should not be used to determine the size of `fs_map.c` in `fs.c`.  Rather that should be defined by a Kconfig variable based on the number of file system types required by an application.

**Review by pabigot — COMMENTED** [review]
Just a couple comments to start.  While looking at this I see problems with the `fs` implementation that I think should be fixed, so that the test can be made more generic.  I'm going to pause to do a few of those changes before continuing.

**Comment by pabigot** [comment]
I may be mistaken, but I think there's some confusion between mount points (file systems) and file system types (implementations of the `fs` API).  I've added #25883 to enable support for a custom filesystem implementation as used in this PR.  Supporting that required making some additional changes to the test, which can be viewed at https://github.com/pabigot/zephyr/commits/pr/25570.  In particular, you cannot rely on the `type` field of a mount point to uniquely identify the mount point.

I do think the test should be renamed to `fs_common` since it's not actually testing an implementation of the fs API, but rather the glue code that lives between the functions the user calls and a specific fs implementation.

This will be a valuable addition to Zephyr's test infrastructure, but there is significant rework to be done.  The test should be using `zassert()` macros throughout, rather than returning `TC_FAIL` which doesn't allow a maintainer to see what specifically wasn't satisfied.  Some of my patches referenced above add examples.

Also there should be infrastructure to make sure that the file system registry, which is owned by the glue code, is reset to a known state after each test completes.  That can probably be done by using setup/teardown functions through [this api](file:///mnt/devel/external/zp/zephyr/doc/_build/html/guides/test/ztest.html#c.ztest_unit_test_setup_teardown).

**Comment by mengxianglinx** [comment]
@zephyrbot Thank you for your concern. I will refine my code according to your #25883 .

File: tests/subsys/fs/fs_api/src/test_fs.c [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+static char buffer[BUF_LEN];
+static char *read_pos = buffer;
+static char *cur = buffer;
+static int file_length;
+static struct fs_mount_t *mp[FS_TYPE_END];
+static bool nospace;
+
+static int temp_open(struct fs_file_t *zfp, const char *file_name)
+{
+	if (zfp && zfp->filep) {
```
Comment (sjg20) [inline tests/subsys/fs/fs_api/src/test_fs.c:27]: Why check zfp if you don't check it below?

File: tests/subsys/fs/fs_api/src/test_fs.c [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+		if (strcmp(zfp->filep, file_name) == 0) {
+			/* file has been opened */
+			return -EIO;
```
Comment (sjg20) [inline tests/subsys/fs/fs_api/src/test_fs.c:30]: EEXIST?

**Review by sjg20 — APPROVED** [review]

File: tests/subsys/fs/fs_api/src/test_fs.c [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+	return 0;
+}
+
+static int temp_statvfs(struct fs_mount_t *mountp,
+			 const char *path, struct fs_statvfs *stat)
+{
+	if (stat == NULL) {
+		return -EINVAL;
+	}
+
```
[the remark this answers is not in the document]
Comment (mengxianglinx) [reply tests/subsys/fs/fs_api/src/test_fs.c:289]: Thank you @pabigot , updated

**Review by pabigot — APPROVED** [review]
Looks good and has nice coverage; thanks.

There will be a collision with #25883 which removes the hard-coded limit on the number of supported file systems.  Patches to the new test to support the changed functionality are in https://github.com/pabigot/zephyr/commits/xpr/25570.  Whichever PR is the second to be merged will need them.

**Review by nashif — APPROVED** [review]

**Pull request merged**